### PR TITLE
MAke eventhub sending fit in 256k size limit

### PR DESCRIPTION
**Pull request by MikeStall**
EventHub can't send more than 256k of data at once. So the EventHub async collector needs to batch it. 
This is generally useful, so make the collector be public. 

Fix https://github.com/Azure/azure-webjobs-sdk/issues/834

Need to think of a good way to test this - none of the EventHub methods are virtual.

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs [hunk 1]
```diff
@@ -1,50 +1,93 @@
...
+                {
+                    var size = (int)item.SerializedSizeInBytes;
 
-            if (flush)
-            {
-                await this.FlushAsync(cancellationToken);
+                    if (size > MaxByteSize)
+                    {
+                        // Single event is too large to add.
+                        string msg = string.Format("Event is too large. Event is approximately {0}b and max size is {0}b", size, MaxByteSize);
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs]: I think your substitution indices are off - both are {0}

Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs]: Let's add tests for this

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs [hunk 2]
```diff
@@ -1,50 +1,93 @@
...
+                        throw new InvalidOperationException(msg);
+                    }
+
+                    if ((_currentByteSize + size > MaxByteSize) || (_list.Count >= BatchSize))
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs]: Can we just get rid of this empty if by applying Demorgans to it?

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs [hunk 3]
```diff
@@ -1,50 +1,93 @@
...
-    // Core object to send events to EventHub. 
-    // Any user parameter that sends events will eventually get bound to this object. 
-    // This gets wrappers with various adapters and passed to the user function. 
-    internal class EventHubAsyncCollector : IAsyncCollector<EventData>
+    /// <summary>
+    /// Core object to send events to EventHub. 
+    /// Any user parameter that sends EventHub events will eventually get bound to this object. 
+    /// This will queue events and send in batches, also keeping under the 256kb event hub limit per batch. 
+    /// </summary>
+    public class EventHubAsyncCollector : IAsyncCollector<EventData>
```
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs]: It doesn't *need* to be public. But I made it public since it's useful in standalone cases. It's also useful for testing a WebJobs function where I want to directly invoke the C# method outside of the model binder.  

The public surface tests don't appear to include the service bus extension.

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs (hunk 2 above)
Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs]: Yes.  I thought it'd be easier to follow the logic like this, but I can flip it.

File: src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs (hunk 3 above)
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs]: I just pushed public surface tests for ServiceBus and your new Logging. Please pull and rebase. You should also review the public types in the Logging assembly, now that this test exposes the list. See commit https://github.com/Azure/azure-webjobs-sdk/commit/af8406f9faeebbd5649fa9038ace3c257cf41227

Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs]: Awesome - I'll rebase on those.  I'd still like to make it public.

**Comment by mathewc** [comment]
:ship:
